api/v1/common_types.go

```diff
@@ -6,6 +6,26 @@ type Image struct {
 	Git *GitSource `json:"git,omitempty"`
 	// Name of container image (example: "docker.io/your-username/your-image").
 	Name string `json:"name,omitempty"`
+	// Upload is a signal that a local tar of the directory should be uploaded to be built as an image.
+	Upload *UploadSpec `json:"upload,omitempty"`
+}
+
+type UploadSpec struct {
+	// Md5Checksum is the md5 checksum of the tar'd repo root requested to be uploaded and built.
+	// +kubebuilder:validation:MaxLength=32
+	// +kubebuilder:validation:MinLength=32
+	// +kubebuilder:validation:Pattern="^[a-fA-F0-9]{32}$"
+	Md5Checksum string `json:"md5checksum,omitempty"`
+}
+
+type UploadStatus struct {
```

Comment:
`type ImageStatus` ... Datasets will likely support a separate upload functionality for the data itself.